pkg/snapshot/plugin.go

```diff
@@ -131,17 +140,182 @@ func (p *Plugin) SetVolumeID(unstructuredPV runtime.Unstructured, volumeID strin
 	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstructuredPV.UnstructuredContent(), pv); err != nil {
 		return nil, errors.WithStack(err)
 	}
-
-	if pv.Spec.PortworxVolume == nil {
-		return nil, errors.New("spec.portworxVolume not found")
+	if pv.Spec.CSI != nil {
+		// PV is provisioned by CSI driver
+		driver := pv.Spec.CSI.Driver
+		if driver == pxdDriverName {
+			pv.Spec.CSI.VolumeHandle = volumeID
+		} else {
+			return nil, fmt.Errorf("unable to handle CSI driver: %s", driver)
+		}
+	} else if pv.Spec.PortworxVolume != nil {
+		// PV is provisioned by in-tree driver
+		pv.Spec.PortworxVolume.VolumeID = volumeID
+	} else {
+		return nil, errors.New("spec.csi and spec.portworxVolume not found")
 	}
 
-	pv.Spec.PortworxVolume.VolumeID = volumeID
-
 	res, err := runtime.DefaultUnstructuredConverter.ToUnstructured(pv)
 	if err != nil {
 		return nil, errors.WithStack(err)
 	}
 
 	return &unstructured.Unstructured{Object: res}, nil
 }
+
+func (p *Plugin) getVolumeDriver() (volume.VolumeDriver, error) {
+	if len(p.jwtSharedSecret) != 0 {
+		claims := &auth.Claims{
+			Issuer: p.jwtIssuer,
+			Name:   "Stork",
+
+			// Unique id for stork
+			// this id must be unique across all accounts accessing the px system
+			Subject: p.jwtIssuer + "." + uniqueID,
+
+			// Only allow certain calls
+			Roles: []string{"system.user"},
```

Comment:
I'm not sure whether we will be able to initiate Cloud-Snapshot using `system.user` , lets confirm this once on px-security enabled cluster

Reply:
I think it will allow but I am yet to double check. This is exactly similar to how stork will handle tokens.